content/advent-2018/observability-in-the-kitchen.md

```diff
@@ -0,0 +1,204 @@
+```
+
+### Configuring the Prometheus Server
+Before starting the server, we need to configure it.  A yaml file (`prometheus.yml`) is used to specify things such as what endpoints to scrape and how frequently to scrape it. 
+
+```
+scrape_configs:
+  - job_name: 'bme280'
+    scrape_interval: 10s
+
+    static_configs:
+    - targets: 
+      - '192.168.0.100:8080'
+      - '192.168.0.101:8080'
+```
+
+### Running the Exporter on the Raspberry Pi
+With the server and the custom exporter running, Prometheus will come over the network to scrape the metrics exported from the exporter running on the Raspberry Pi that is connected to the temperature and humidity sensor. 
+
+It would be nice to have my custom exporter start automatically upon powering up the Raspberry Pi and there are a couple of ways to enable this. I decided to create a systemd unit since it seems to be the most flexible way to manage services that run in the background on a Linux system. It allows you to run services before the desktop environment starts, wait until you have access to other processes (e.g. networking, graphical desktop), or simply restart your program over and over again until it works. 
+
+```
+[Unit]
+Description=Sourdough Prometheus Exporter
+After=network.target
+
+[Install]
+WantedBy=multi-user.target
+
+[Service]
+WorkingDirectory=/home/pi/go/src/github.com/sourdough-prometheus-exporter
+ExecStart=/home/pi/go/src/github.com/sourdough-prometheus-exporter/sourdough-prometheus-exporter
+User=pi
+Restart=always
+```
+
+### Metrics Exposition
+When I run the exporter and go to the `/metrics` endpoint, I see something like this:
+
+```
+# HELP bme280_humidity Humidity in percentage of relative humidity
+# TYPE bme280_humidity gauge
+bme280_humidity 49.532752990722656
+
+# HELP bme280_temperature_celsius Temperature in celsius degree
+# TYPE bme280_temperature_celsius gauge
+bme280_temperature_celsius 21.880083084106445
+```
+
+The exposition is in a line-by-line text-based [format](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md) typically with a HELP and TYPE comment line for each metric.  
+
+Just because metrics can be scraped does not mean that the format is compliant with the [standards](https://prometheus.io/docs/practices/naming/#metric-names).  You can run the `promtool` command-line utility that comes with Prometheus and use it to perform link checks on your metrics for consistency and correctness, by passing the metrics over stdin. 
+
+Example:  `curl http://localhost:8080/metrics | promtool check metrics`
+
+
+### Dashboarding, PromQL, Alertmanager
+Once the scraped data is stored in the time-series database, we can use it to create dashboards. Grafana is a popular choice and has support for querying Prometheus.  All you have to do is create a Prometheus data source in Grafana and you can start creating graphs by querying your scraped data.
+
+![Example Graph](https://cdn.pbrd.co/images/HQO4X2T.jpg)
```

Comment:
Completely out of the Go scope, but I'm curious about the Air Pressure units (because it's been a long time since I last touched it!): 
Googled reminded me that the air pressure at sea level is 101.325 kPa or 1.013 bar
The values in the Air Pressure graph show ca. 1.01 K - which did you mean here?

Reply:
I looked into this and wondered if maybe the sensor was saying 1.01K millibars or maybe it has something to do with needing to provide the local altitude first.  Then I looked at readFloatPressure() here and wasn't sure anymore: https://learn.sparkfun.com/tutorials/sparkfun-bme280-breakout-hookup-guide/all#fun.  I decided to just change the pic to exclude the Air Pressure graph for now since I didn't talk about it in the article anyways.